### DISCUSSION TITLE
Request for Fallback Colors

### DISCUSSION THREAD

**Discussion by barrycenter** [tailwindlabs/tailwindcss#10701] [Ideas]
Many of the newer CSS color functions, such as `lch()` or `color()`, supported in Safari 15+ and Chrome 111+, require fallback colors to display properly in all other browsers. 

Normally, you'd do this in your css like this: 
```css
.success {
    color: #19FC00; /* fallback color */
    color: lch(87% 132 134);
}
.brand-primary {
    color: rgb(0% 96.63% 97.67%); /* fallback color */
    color: lch(88% 132 198);
}
```

It would be nice if we could easily specify our fallback colors when defining a color in `tailwind.config.js`. 

Perhaps by providing an array of color definitions, when needed, like this:
```js
module.exports = {
  theme: {
    extend: {
      colors: {
        success: [
          "#19FC00", // fallback color
          "lch(87% 132 134 / <alpha-value>)"
        ],
       "brand-primary": [
          "rgb(0% 96.63% 97.67% / <alpha-value>)", // fallback color
          "lch(88% 132 198 / <alpha-value>)"
        ],
       "warning": [
          "rgb(var(--color-warning-rgb) / <alpha-value>)", // fallback color
          "lch(var(--color-warning-lch) / <alpha-value>)"
        ],
        "bright-purple": {
           ...
           400: [
             "rgb(88.06% 33.88% 100% / <alpha-value>)",  // fallback color
             "color(display-p3 0.8335 0.3474 1 / <alpha-value>)"
           ],
           500: [
             "rgb(82.31% 0% 98.23% / <alpha-value>)",  // fallback color
             "color(display-p3 0.7759 0 0.9926 / <alpha-value>)"
           ],
           600: [
             "rgb(66.48% 0% 79.53% / <alpha-value>)",  // fallback color
             "color(display-p3 0.6261 0 0.8037 / <alpha-value>)"
           ],
           ...
        }        
      }
    }
  }
}
```

This is simply adding an optional array of colors, in addition to the normal string definitions, so all the existing definitions in users' `tailwind.config.js` will still be completely valid. It's just adding an additional mechanism to give users the ability to actually specify fallback colors which is sorely needed in Tailwind.

### Modern CSS Colors

* Here's an [LCH Color Picker](https://css.land/lch/) which will allow you to play with LCH colors if you have Safari (Mac or iOS) or a version of Chrome which supports the `color()` function. 
* You can read more about modern CSS colors in this article: [LCH colors in CSS: what, why, and how?](https://lea.verou.me/2020/04/lch-colors-in-css-what-why-and-how/)

**Comment by robdekort** [comment]
This would be **such** a welcome addition. Other methodologies to currently use spanking colours feel very clunky. I've done stuff like this in the past when using the P3 color function: `bg-primary supports-[background-color:color(display-p3)]:bg-primary-p3`.

**Comment by robdekort** [comment]
Hey @barrycenter I just found out about a super easy way to use fallback colors in Tailwind. I wrote a quick post about it: https://1902.studio/en/journal/using-oklch-colors-in-tailwind-css

**Reply by robdekort** [reply]
One thing I found is that referencing colours unfortunately doesn't work anymore using this technique. For example when customizing TW typography: `'--tw-prose-quote-borders': theme('colors.primary.DEFAULT')`. It compiles the color with `<alpha-value>` included.  Other then setting another color value without the alpha-value I haven't found a workaround for this. Curious if anyone has any thoughts.

**Reply by danielwaltz** [reply]
Not sure if it actually helps in your case, but I ran into this issue using [UnoCSS](https://unocss.dev) for colors and your postcss solution worked like a dream for me (thank you so much)!

I have not run into anything similar to what you mention above, but I believe it's because UnoCSS doesn't require adding an `<alpha-value>` anywhere in the color definitions as transparency is automatically/dynamically handled by UnoCSS.

**Reply by robdekort** [reply]
Glad it helped! I hadn’t heard of UnoCSS before. Some striking similarities there :-).

**Comment by man-trackunit** [comment]
Check out Oxide from the Tailwind Connect [video](https://m.youtube.com/watch?v=CLkxRnRQtDE%2F51%3A30) (around 51:30)

**Reply by robdekort** [reply]
Yeah I just noticed it last week. Good news. One thing I hope they implement is to also auto generate color(display-p3) definitions so older versions of Safari going back to 2017 can enjoy richer colors. The postcss plugin I'm currently using does this (optionally).
